lock-impl/src/test/java/com/palantir/lock/impl/ThreadInfoLockServiceImplTest.java

```diff
@@ -0,0 +1,308 @@
+    }
+
+    @Test
+    public void recordsThreadInfoForUnlock() throws InterruptedException {
+        LockResponse response =
+                lockService.lockWithFullLockResponse(LockClient.ANONYMOUS, LOCK_1_THREAD_1_WRITE_REQUEST);
+        forceSnapshot();
+
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_1)).isEqualTo(ANONYMOUS_TEST_THREAD_1);
+
+        lockService.unlock(response.getToken());
+        forceSnapshot();
+
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_1)).isNull();
+    }
+
+    @Test
+    public void recordsThreadInfoForMultipleLocks() throws InterruptedException {
+        int numThreads = 10;
+        int numLocksPerThread = 10;
+
+        List<String> threadNames =
+                IntStream.range(0, numThreads).mapToObj(i -> "test-thread-" + i).collect(Collectors.toList());
+        Map<String, Set<LockDescriptor>> locksPerThread = new HashMap<>();
+        for (String threadName : threadNames) {
+            Set<LockDescriptor> locks = IntStream.range(0, numLocksPerThread)
+                    .mapToObj(i -> StringLockDescriptor.of("test-lock-" + i + "-from-thread-" + threadName))
+                    .collect(Collectors.toSet());
+            locksPerThread.put(threadName, locks);
+        }
+
+        for (String threadName : threadNames) {
+            LockRequest lockRequest = LockRequest.builder(
+                            ImmutableSortedMap.copyOf(locksPerThread.get(threadName).stream()
+                                    .collect(Collectors.toMap(lock -> lock, lock -> LockMode.WRITE))))
+                    .withCreatingThreadName(threadName)
+                    .build();
+            lockService.lockWithFullLockResponse(LockClient.ANONYMOUS, lockRequest);
+        }
+
+        forceSnapshot();
+
+        locksPerThread.forEach((threadName, lockDescriptors) ->
+                lockDescriptors.forEach(lock -> assertThat(getLatestThreadInfoForLock(lock))
+                        .isEqualTo(LockClientAndThread.of(LockClient.ANONYMOUS, threadName))));
+    }
+
+    @Test
+    public void recordsThreadInfoForSharedLock() throws InterruptedException {
+        LockRequest lockRequest1 = LockRequest.builder(ImmutableSortedMap.of(TEST_LOCK_1, LockMode.READ))
+                .withCreatingThreadName(TEST_THREAD_1)
+                .build();
+        LockRequest lockRequest2 = LockRequest.builder(ImmutableSortedMap.of(TEST_LOCK_1, LockMode.READ))
+                .doNotBlock()
+                .lockAsManyAsPossible()
+                .withCreatingThreadName(TEST_THREAD_2)
+                .build();
+
+        LockResponse lockResponse = lockService.lockWithFullLockResponse(LockClient.ANONYMOUS, lockRequest1);
+        lockService.lockWithFullLockResponse(LockClient.ANONYMOUS, lockRequest2);
+
+        forceSnapshot();
+
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_1))
+                .isIn(ANONYMOUS_TEST_THREAD_1, LockClientAndThread.of(LockClient.ANONYMOUS, TEST_THREAD_2));
```

Comment:
The reason for giving two options is because it is undefined which ClientThread takes precedence in case a lock is held by multiple threads. This boils down to iteration order of `heldLoksTokenMap`, which is arbitrary (I think we are using ConcurrentHashMap under the hood).
Asserting that the client thread is one of two options is the strongest assertion we can make here IMO.